### ISSUE TITLE
Some close lines do not render properly on version 0.15.0 (and master)

### ISSUE THREAD

**Issue by WeebDataHoarder** [libass/libass#497]
After testing [libass 0.15.0](https://github.com/libass/libass/releases/tag/0.15.0) it was noticed that several previously working subtitles rendered incorrectly. Whole lines are dropped (or rendered elsewhere), debug logs don't show much.

A bisect between 0.14.0 and 0.15.0 (also done between 0.14.0 and master) brought d4728d70e74707ffbd5ff01cb2d543fb450b6494 as the first bad commit. Testing from the commit just before results in normal working subtitles.

The lines usually have tight margins and are close to each other.

Here is a comparison of lines and their renders.

| Time | Good result https://github.com/libass/libass/commit/86114ad2d532f83496ebd161f6538cf184dd21de / 0.14.0 | Bad result d4728d70e74707ffbd5ff01cb2d543fb450b6494 / master |
| :--- | :---: | :---: |
|`00:59.00` | ![good_1](https://user-images.githubusercontent.com/57538841/115760480-b49a3800-a3a1-11eb-9252-f8fa82af23c8.png) | ![bad_1](https://user-images.githubusercontent.com/57538841/115761296-7ea98380-a3a2-11eb-89b7-500dbbff1dcd.png) | 
| `00:15.00` |  ![good_2](https://user-images.githubusercontent.com/57538841/115760506-b6fc9200-a3a1-11eb-818d-2367279c9ca5.png) | ![bad_2](https://user-images.githubusercontent.com/57538841/115761325-86692800-a3a2-11eb-912d-4a9781761c21.png) | 
| `00:47.00` | ![good_3](https://user-images.githubusercontent.com/57538841/115760522-b82dbf00-a3a1-11eb-9f9a-7f58d6b5d95e.png) | ![bad_3](https://user-images.githubusercontent.com/57538841/115761335-8832eb80-a3a2-11eb-893a-672aee3a877d.png) | 

Here's the .ass file on a [lyrics.zip](https://github.com/libass/libass/files/6359731/lyrics.zip) zip. Fonts are attached for convenience, but also produce similar results when loading fonts directly.

I'm also available on IRC to discuss this issue.

**Comment by WeebDataHoarder** [comment]
These images were rendered from xy-VSFilter using a canvas of size 2048x192 (same aspect ratio, different video canvas). Do note some of the lines change ordering vs what libass renders (while still having the same behavior of "cutting out" whole lines).

![](https://mei.tuuba.moe/jDuSMj.png)
![](https://mei.tuuba.moe/Z0MELP.png)
![](https://mei.tuuba.moe/YuX1so.png)

**Comment by TheOneric** [comment]
To recap all findings from IRC:

 - implicit positioning changing slightly by the bisected commit is expected, VSFilter compatible and correct
   - Using implicit positioning when high positon precision is required is a bit risky if anything influencing placement isn't part of the guaranteed (mostly) stable VSFilter compatible behaviour.
 - xy-VSFilter also only shows one line per frame *(indicating that the new behaviour is mostly correct)*
 - when no seeking/skipping ahead is involved both xy-VSF and libass-after-d4728d7 always only show the romaji lines (top)
   - with seeking sometimes the English (bottom) line can show up.
     This is because of the in-file ordering of the lines which is either `romaji_part1 - english - romaji_part2` or `romjai - english`.
     On time-and-space collision the line which comes first in data gets to keep its base state and once placed lines are frozen in place. In case of the first ordering and    seeking from a positon _before_ `romaji_part1` to a position _after_ `romaji_part1`, omaji_part1` isn't processed and so The English line gets to remain in its base state.
     This ordering is brittle in non-continuous scenarios.
   

And as to why the whole line disappears, instead of being hidden partially, mpv visualises why, cyan is video area, black outside of video area *(not sure if rendering subs outside of video area is a mpv *bug* or *feature*, but it helps here)*:
![visual](https://user-images.githubusercontent.com/1668471/115935096-be4f9880-a492-11eb-96e6-a55bff146434.png)

Romaji lines use `\an8` and English ones use `\an2`. `\an2` text is moved **upwards** on collision, which in this case pushes the whole event out of the video area. With how tight the text was fitted to canvas minus margins, after `d4728d7` all simultaneous lines collide. If both lines had the same alignment, the shift would have been more subtle pushing one line per frame only *slightly* out of margin bounds instead of completely out of video area *(possibly the text would even have left the video)*.
 
tl;dr: everything here works as expected from ASS and is VSF-compatible

**Comment by astiob** [comment]
> not sure if rendering subs outside of video area is a mpv _bug_ or _feature_

It’s libass’s “margins” feature. For ASS in mpv, it is toggled by `--sub-ass-force-margins`, which I believe is _disabled_ by default. (The same is enabled by default for non-ASS subtitles, even though mpv internally converts them to ASS in order to render them via libass.) Check your mpv options: if you don’t explicitly enable this, this shouldn’t happen.

**Comment by TheOneric** [comment]
> > not sure if rendering subs outside of video area is a mpv _bug_ or _feature_
> 
> It’s libass’s “margins” feature. For ASS in mpv, it is toggled by `--sub-ass-force-margins`, which I believe is _disabled_ by default. […] Check your mpv options: if you don’t explicitly enable this, this shouldn’t happen.

I tested with mpv 0.29 and my normal config, which does not enable `--sub-ass-force-margins` and mpv-git-master without any config file, both showed the above behaviour.
When I *enable* `--sub-ass-force-margins`, subs are placed as if the whole player canvas was the video-area:
![mpv_ass-force-margins](https://user-images.githubusercontent.com/1668471/116000190-742ff980-a5ef-11eb-9e76-81332c5b1702.png)

**Comment by astiob** [comment]
Cool, that sounds like a bug to me. Is this a regression in 450b791d9847e1730a3024d07f46c0b3d69d75d7 or 4698e1474a573659cfe97ebde3335ef4f4c48237?

**Comment by TheOneric** [comment]
Checking out  d4728d7 and then reverting 450b791 or 4698e14 didn't change the result.

I suspect #121 may be responsible for this *(I also edited this into my previous comment just moments before you posted, not sure if you noticed it before)*.
